Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,322 @@
+        // so we need to skip them
+        int changesAhead = 0;
+
+        protected virtual void SerializeKey(NetworkWriter writer, B item) { }
+        protected virtual void SerializeItem(NetworkWriter writer, T item) { }
+        protected virtual B DeserializeKey(NetworkReader reader) => default(B);
+        protected virtual T DeserializeItem(NetworkReader reader) => default(T);
+
+        public bool IsDirty => Changes.Count > 0;
+
+        public ICollection<B> Keys => m_Objects.Keys;
+
+        public ICollection<T> Values => m_Objects.Values;
+
+        // throw away all the changes
+        // this should be called after a successfull sync
+        public void Flush() => Changes.Clear();
+
+        public SyncDictionary()
+        {
+            m_Objects = new Dictionary<B, T>();
+        }
+
+        public SyncDictionary(IEqualityComparer<B> eq)
+        {
+            m_Objects = new Dictionary<B, T>(eq);
+        }
+
+        void AddOperation(Operation op, B key, T item)
+        {
+            if (IsReadOnly)
+            {
+                throw new System.InvalidOperationException("Synclists can only be modified at the server");
+            }
+
+            Change change = new Change
+            {
+                operation = op,
+                key = key,
+                item = item
+            };
+
+            Changes.Add(change);
+
+            Callback?.Invoke(op, key, item);
+        }
+
+        public void OnSerializeAll(NetworkWriter writer)
+        {
+            // if init,  write the full list content
+            writer.Write(m_Objects.Count);
+
+            foreach (KeyValuePair<B, T> syncItem in m_Objects)
+            {
+                SerializeKey(writer, syncItem.Key);
+                SerializeItem(writer, syncItem.Value);
+            }
+
+            // all changes have been applied already
+            // thus the client will need to skip all the pending changes
+            // or they would be applied again.
+            // So we write how many changes are pending
+            writer.Write(Changes.Count);
+        }
+
+        public void OnSerializeDelta(NetworkWriter writer)
+        {
+            // write all the queued up changes
+            writer.Write(Changes.Count);
+
+            for (int i = 0; i < Changes.Count; i++)
+            {
+                Change change = Changes[i];
+                writer.Write((byte)change.operation);
+
+                switch (change.operation)
+                {
+                    case Operation.OP_ADD:
+                        SerializeKey(writer, change.key);
+                        SerializeItem(writer, change.item);
+                        break;
+
+                    case Operation.OP_CLEAR:
+                        break;
+
+                    case Operation.OP_REMOVE:
+                        SerializeKey(writer, change.key);
+                        SerializeItem(writer, change.item);
+                        break;
```

Comment:
Don't need to serialize the item value— the key is enough to identify what to remove.

Reply:
This is necessary for some devs in the Callback - merged with all the other cases